### PR TITLE
treewide: Allow the linker to be chosen independently

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

This will begin the process of breaking up the `useLLVM` monolith. That
is good in general, but I hope will be good for NetBSD and Darwin in
particular.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 122778` at 168194cc run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>38 packages skipped due to time constraints:</summary>
  <ul>
    <li>adoptopenjdk-icedtea-web</li>
    <li>chromium</li>
    <li>clickhouse</li>
    <li>firefox-esr-78-unwrapped (firefoxPackages.firefox-esr-78)</li>
    <li>xulrunner (firefox-unwrapped)</li>
    <li>llvmPackages_9.clangNoCompilerRt (haskellPackages.llvmPackages.clangNoCompilerRt)</li>
    <li>llvmPackages_9.clangNoCompilerRtWithLibc (haskellPackages.llvmPackages.clangNoCompilerRtWithLibc)</li>
    <li>llvmPackages_9.clangNoLibc (haskellPackages.llvmPackages.clangNoLibc)</li>
    <li>llvmPackages_9.clangNoLibcxx (haskellPackages.llvmPackages.clangNoLibcxx)</li>
    <li>llvmPackages_9.clangUseLLVM (haskellPackages.llvmPackages.clangUseLLVM)</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>3 packages built successfully:</summary>
  <ul>
    <li>bintools</li>
    <li>bintools-unwrapped</li>
    <li>bintoolsNoLibc</li>
  </ul>
</details>

**Comment by mweinelt** [comment]
This breaks the firefox build on x86_64-linux.

```
 0:09.91(B checking for linker...(B
 0:09.92(B DEBUG: Executing: `/nix/store/l9fklk5dj4b4gzpc28d16ckfq5y1i1xk-clang-wrapper-11.1.0/bin/clang -std=gnu99 -fuse-ld=lld -Wl,--version`(B
 0:09.92(B ERROR: Could not use lld as linker(B
Error running mach:

    ['configure', '--prefix=/nix/store/dr9fplxh68mc0314d1x8fvrz1ggvkvln-firefox-unwrapped-88.0.1', '--enable-application=browser', '--with-system-jpeg', '--with-system-zlib', '--with-system-libevent', '--with-system-libvpx', '--with-system-png', '--with-system-icu', '--enable-system-ffi', '--enable-system-pixman', '--disable-tests', '--disable-necko-wifi', '--disable-updater', '--enable-jemalloc', '--enable-default-toolkit=cairo-gtk3-wayland', '--with-libclang-path=/nix/store/jlzl1c2w1d884ykzlmpby102y1rdzzw3-clang-11.1.0-lib/lib', '--with-system-nspr', '--with-system-nss', '--enable-lto', '--disable-elf-hack', '--enable-linker=lld', '--enable-alsa', '--enable-pulseaudio', '--enable-ffmpeg', '--enable-negotiateauth', '--enable-webrtc', '--disable-crashreporter', '--disable-debug', '--enable-optimize', '--enable-strip', '--enable-release', '--enable-official-branding', '--with-google-location-service-api-keyfile=/build/ga', '--with-google-safebrowsing-api-keyfile=/build/ga']

The error occurred in code that was called by the mach command. This is either
a bug in the called code itself or in the way that mach is calling it.
You can invoke |./mach busted| to check if this issue is already on file. If it
isn't, please use |./mach busted file configure| to report it. If |./mach busted| is
misbehaving, you can also inspect the dependencies of bug 1543241.

If filing a bug, please include the full output of mach, including this error
message.

The details of the failure are as follows:

Exception: Process executed with non-0 exit code 1: ['/nix/store/pxhyfgz2c7y5n35hjphdi0ady7xdl4sm-python3-3.8.9/bin/python3', '/build/firefox-88.0.1/configure.py', '--prefix=/nix/store/dr9fplxh68mc0314d1x8fvrz1ggvkvln-firefox-unwrapped-88.0.1', '--enable-application=browser', '--with-system-jpeg', '--with-system-zlib', '--with-system-libevent', '--with-system-libvpx', '--with-system-png', '--with-system-icu', '--enable-system-ffi', '--enable-system-pixman', '--disable-tests', '--disable-necko-wifi', '--disable-updater', '--enable-jemalloc', '--enable-default-toolkit=cairo-gtk3-wayland', '--with-libclang-path=/nix/store/jlzl1c2w1d884ykzlmpby102y1rdzzw3-clang-11.1.0-lib/lib', '--with-system-nspr', '--with-system-nss', '--enable-lto', '--disable-elf-hack', '--enable-linker=lld', '--enable-alsa', '--enable-pulseaudio', '--enable-ffmpeg', '--enable-negotiateauth', '--enable-webrtc', '--disable-crashreporter', '--disable-debug', '--enable-optimize', '--enable-strip', '--enable-release', '--enable-official-branding', '--with-google-location-service-api-keyfile=/build/ga', '--with-google-safebrowsing-api-keyfile=/build/ga']

  File "/build/firefox-88.0.1/python/mozbuild/mozbuild/build_commands.py", line 177, in configure
    return driver.configure(
  File "/build/firefox-88.0.1/python/mozbuild/mozbuild/controller/building.py", line 1529, in configure
    status = self._run_command_in_objdir(
  File "/build/firefox-88.0.1/python/mozbuild/mozbuild/base.py", line 897, in _run_command_in_objdir
    return self.run_process(cwd=self.topobjdir, **args)
  File "/build/firefox-88.0.1/python/mach/mach/mixin/process.py", line 176, in run_process
    raise Exception(
builder for '/nix/store/lq1xykqwb7nz6n79ax480kwshggw82lh-firefox-unwrapped-88.0.1.drv' failed with exit code 1
```

**Comment by Ericson2314** [comment]
Yeah i found I think workarounds for everything but Firefox. Hmmm.

**Comment by Ericson2314** [comment]
Firefox now has a (somewhat ugly) workaround. I don't really like the override, but writing down every possible combination is also hard. That's when I wish we had a relational language / solver :D.

**Comment by Ericson2314** [comment]
@matthewbauer It's been a minute, but I'd love for you to way on this. I first mentioned doing something like this quite some time ago :).

**Comment by Ericson2314** [comment]
OK the fix for firefox is a lot less ugly now.

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 1]
```diff
@@ -98,16 +98,22 @@ let
   # clang LTO on Darwin is broken so the stdenv is not being changed.
   # Target the LLVM version that rustc -Vv reports it is built with for LTO.
   # rustPackages_1_45 -> LLVM 10, rustPackages -> LLVM 11
-  llvmPackages = if stdenv.isDarwin
-                 then buildPackages.llvmPackages
-                 else if lib.versionAtLeast rustc.llvm.version "11"
-                      then buildPackages.llvmPackages_11
-                      else buildPackages.llvmPackages_10;
+  llvmPackages0 =
+    /**/ if stdenv.isDarwin
```
Comment (lovesegfault) [inline pkgs/applications/networking/browsers/firefox/common.nix:102]: What's the point of the `/**/` here?

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 2]
```diff
@@ -98,16 +98,22 @@ let
   # clang LTO on Darwin is broken so the stdenv is not being changed.
   # Target the LLVM version that rustc -Vv reports it is built with for LTO.
   # rustPackages_1_45 -> LLVM 10, rustPackages -> LLVM 11
-  llvmPackages = if stdenv.isDarwin
-                 then buildPackages.llvmPackages
-                 else if lib.versionAtLeast rustc.llvm.version "11"
-                      then buildPackages.llvmPackages_11
-                      else buildPackages.llvmPackages_10;
+  llvmPackages0 =
```
Comment (lovesegfault) [inline pkgs/applications/networking/browsers/firefox/common.nix:101]: Could you name this `llvmPackages'`?

**Review by lovesegfault — APPROVED** [review]
Left two small nits, but the changes look good overall to me. I suspect it may be worth it to do a separate hydra run of this just to be sure, even if this is going into staging and not master. Up to the author.

**Review by sternenseemann — APPROVED** [review]
LGTM. Not sure if `bootBintools` is the clearest terminology in this case, but OTOH I am also unable to express what I think “boot” is supposed to mean in that context.

File: pkgs/applications/networking/browsers/firefox/common.nix (hunk 1 above)
Comment (Ericson2314) [reply pkgs/applications/networking/browsers/firefox/common.nix:102]: Just lining things up :).

**Comment by Ericson2314** [comment]
@sternenseemann Yeah I don't really like the name either, but I don't intend for it to be really stable so we can just change it later. I somewhat hope as the bootstrapping logic becomes more standardized across all the combinations, we can move it out of the default.nix anyways.

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -48,7 +48,22 @@ in
       lib.mapNullable (rs: rs ++ [ bintools ]) (stdenv.allowedRequisites or null);
   };
 
-  stdenvNoLibs = mkStdenvNoLibs stdenv;
+  stdenvNoLibs =
+    if stdenv.hostPlatform != stdenv.buildPlatform && (stdenv.hostPlatform.isDarwin || stdenv.hostPlatform.isDarwin.useLLVM or false)
```
Comment (Artturin) [inline pkgs/top-level/all-packages.nix:52]: `stdenv.hostPlatform.isDarwin.useLLVM or false``stdenv.hostPlatform.isDarwin.useLLVM or false`

should be

`stdenv.hostPlatform.isDarwin.useLLVM or false``stdenv.hostPlatform.useLLVM or false`

```
nix-repl> pkgsCross.x86_64-freebsd.stdenv.hostPlatform.useLLVM
true
```

Comment (Ericson2314) [reply pkgs/top-level/all-packages.nix:52]: Oops! Thanks for catching!

Comment (Artturin) [reply pkgs/top-level/all-packages.nix:52]: https://github.com/NixOS/nixpkgs/pull/274047

**Pull request merged**